src/impl_/coroutine.rs

```diff
@@ -1,19 +1,33 @@
-use crate::coroutine::Coroutine;
-use crate::impl_::wrap::OkWrap;
-use crate::{IntoPy, PyErr, PyObject, Python};
 use std::future::Future;
 
-/// Used to wrap the result of async `#[pyfunction]` and `#[pymethods]`.
-pub fn wrap_future<F, R, T>(future: F) -> Coroutine
+use crate::{
+    coroutine::Coroutine,
+    types::{PyModule, PyString},
+    IntoPy, Py, PyClass, PyErr, PyObject, Python,
+};
+
+pub fn new_coroutine<F, T, E>(name: Py<PyString>, qualname: Py<PyString>, future: F) -> Coroutine
 where
-    F: Future<Output = R> + Send + 'static,
-    R: OkWrap<T>,
+    F: Future<Output = Result<T, E>> + Send + 'static,
     T: IntoPy<PyObject>,
-    PyErr: From<R::Error>,
+    PyErr: From<E>,
 {
-    let future = async move {
-        // SAFETY: GIL is acquired when future is polled (see `Coroutine::poll`)
-        future.await.wrap(unsafe { Python::assume_gil_acquired() })
```

Comment:
I've removed this `Python::assume_gil_acquired()` using `OkWrap::wrap_no_gil`. Actually, this modification was not required for this PR (it is for the following about `allow_threads`), but because I had to modified `wrap_future`, I've took the opportunity to do it.